packaging/windows/clisdk/registrykeys.wxs

```diff
@@ -2,12 +2,17 @@
 <Wix xmlns="http://schemas.microsoft.com/wix/2006/wi">
   <?include "Variables.wxi" ?>
   <Fragment>
-    <ComponentGroup Id="AuthoredRegistryKeys">
-      <Component Id="SetupRegistry_x86" Directory="TARGETDIR" Win64="no">
+    <ComponentGroup Id="AuthoredRegistryKeys_x86node">
+      <Component Id="SetupRegistry" Directory="TARGETDIR" Win64="no">
         <RegistryKey Root="HKLM" Key="SOFTWARE\dotnet\Setup\InstalledVersions\$(var.Platform)\sdk">
           <RegistryValue Action="write" Name="$(var.NugetVersion)" Type="integer" Value="1" KeyPath="yes"/>
         </RegistryKey>
       </Component>
+      <Component Id="SDKInstallLocation" Directory="TARGETDIR" Win64="no" Permanent="yes">
```

Comment:
Any reason this is marked as permanent? You're generating stable component GUIDs, so multiple CLI MSIs should just ref count this component correctly, or is there another reason to keep the registry key after all the CLIs have been removed?

Reply:
This idea of permanence comes from Visual Studio - they support one and only one location for shared components for the lifetime of the machine. I feel that marking the registry key as permanent is in alignment with this philosophy.

Reply:
It is not clear to me what this permanent flag does, but I want to know what are its implications for stand-along installs as well. We need to keep both scenarios in mind (vs and stand-along) when making this change.

Reply:
I still have concerns about having a single 32-bit registry key for the install location. As it stands today, installing both the 64-bit and 32-bit SDKs on the same machine to the same shared folder will cause the last install to win and overwrite the contents.

It's also conceivable that the SDKs might ship different files which means you could end up with a mix of 32/64 bit files.

Shouldn't the SDK follow the pattern we had in .NET with the framework and framework64 folders? Also, because the folders can be overwritten, the components will end up writing to the same location, but will have different component IDs because those are generated based on the KeyPath which will like be based off ProgramFilesFolder or ProgramFiles64Folder at the MSI level. 

Here's an example: Microsoft.Common.CrossTargeting.targets

In the 64-bit installer, the component ID is {6922A509-C686-4C00-AB3B-E032F05E1154}
In the 32-bit installer, the component ID is {541E9B5E-4DC1-4DB8-A48A-DDA3225DE41D}

If you install both SDKs and override the default location to something like C:\Shared, you'll end up with the copy of the file from the last installer always because the component wouldn't be viewed as shared. For text files this likely won't cause an overwrite (unless they are different or have different timestamps), but for binaries, the file hashes would be different and since the component is considered not installed, it will be overwritten as there's no reference counting.

Reply:
@livarcocc permanent indicates that the component would never be removed during an uninstall. So in the case of registry keys, an uninstall would typically remove the key (assuming it's not a shared component). In this case, the key is persisted. Once uninstalled, the component is effectively orphaned as there is no associated product with it.

Reply:
The registry key records the installation location of SDK bundles in both scenarios; stand-alone & thru Visual Studio. Marking he registry key as permanent does not impact our stated scenarios.

Reply:
There are a potential of 2 keys written on 64-bit OSs:
x64 SDK installation = HKLM\SOFTWARE\dotnet\Setup\InstalledVersions\**x64**\sdk
x86 SDK installation = HKLM\SOFTWARE\dotnet\Setup\InstalledVersions\**x86**\sdk

An x64 installation should not co-mingle with a x86 installation - this was the case in my limited testing. Furthermore, we should ensure that x64 and x86 installations confine themselves to the SDK bundle installation path stated.

Reply:
Right, the installed versions distinguishes between x86 and x64, but the location doesn't. And since the location serves as the root folder, x86 and x64 will overwrite the files they installs and you'd just have one set of SDK files even though both are installed.

I don't think we're doing anything to stop the co-mingling. Maybe I'm missing something here. With this change you will end up with the SDK in the same location (so it's honoring that part), but if I install everything (32 and 64-bit) to c:\dotnet, I'll end up with both SDKs, but the actual file, e.g. c:\dotnet\sdk\3.0.0\dotnet.dll will be whichever bitness I installed last. If the folder is shared, shouldn't we instead install to dotnet\SDK\3.0.0\x86 and dotnet\SDK\3.0.0\x64?

Reply:
I think it should be [x86/x64] at the front.


```
[custom location]\x86\
    dotnet.exe (x86 launcher)
    shared\ (x86 runtimes)
    sdk\ (x86 sdks)

[custom location]\x86\
    dotnet.exe (x86 launcher)
    shared\ (x86 runtimes)
    sdk\ (x86 sdks)
```

Reply:
Or else we can have two separate custom locations for each bitness, which I think John is saying we do already, and Jacques is saying that is not the case. I don't follow all of the msi specifics.

Reply:
@johnbeisner if this is why you asked me about permanency, I'm actually with @joeloff on this one. Visual Studio Installer treats permanency of the `[SharedInstallDir]` because we install a lot of first- and third-party partner content - including .NET. Some of that is permanent, and much of that can be acquired in many different ways. Many chainers don't use package ref-counting. I added that to WiX-built MSIs first (but they had to take advantage of it) then WiX burn bundles later, and we implemented it in the VS2017 (and newer) installer. But I'm not aware of any other chainers that use it. Since we don't want to break other products/installer, we don't remove shared packages. And because we don't remove shared packages, we can't assume we could ever change `[SharedInstallDir]`.

But this is not the case for MSI components. MSI components are already ref-counted (in fact, my package ref-counting feature borrows much from how MSI did it). So if your various SDK versions use the same component GUIDs to write the location, they all get ref-counted such that the last component will finally remove it (i.e., last one out turns off the lights). So if this location is new going forward, you have an opportunity to better control cleaning this up.

Reply:
Registry key is no longed marked as 'Permanent'